protocol/statetracker/updaters/provider_availability_updater.go

```diff
@@ -0,0 +1,87 @@
+package updaters
+
+import (
+	"context"
+	"sync"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/lavanet/lava/v3/protocol/lavasession"
+	"github.com/lavanet/lava/v3/protocol/metrics"
+	"github.com/lavanet/lava/v3/x/pairing/types"
+	"google.golang.org/grpc"
+)
+
+type ProviderPairingStatusStateQueryInf interface {
+	Providers(ctx context.Context, in *types.QueryProvidersRequest, opts ...grpc.CallOption) (*types.QueryProvidersResponse, error)
+}
+
+type AvailabilityStatus uint64
+
+const (
+	AVAILABLE AvailabilityStatus = iota
+	FROZEN
+)
+
+type ProviderAvailabilityUpdater struct {
+	lock                sync.RWMutex
+	latestEpoch         uint64
+	pairingQueryClient  ProviderPairingStatusStateQueryInf
+	metricsManager      *metrics.ProviderMetricsManager
+	clientCtx           client.Context
+	rpcProviderEndpoint *lavasession.RPCProviderEndpoint
+}
+
+func NewProviderAvailabilityUpdater(
+	stateQuery ProviderPairingStatusStateQueryInf,
+	rpcProviderEndpoint *lavasession.RPCProviderEndpoint,
+	clientCtx client.Context,
+	metricsManager *metrics.ProviderMetricsManager,
+) *ProviderAvailabilityUpdater {
+	return &ProviderAvailabilityUpdater{
+		pairingQueryClient:  stateQuery,
+		clientCtx:           clientCtx,
+		rpcProviderEndpoint: rpcProviderEndpoint,
+		metricsManager:      metricsManager,
+	}
+}
+
+func (pau *ProviderAvailabilityUpdater) UpdateEpoch(epoch uint64) {
+	go pau.runProviderAvailabilityUpdate(epoch)
+}
+
+func (pau *ProviderAvailabilityUpdater) runProviderAvailabilityUpdate(epoch uint64) {
+	pau.lock.Lock()
+	defer pau.lock.Unlock()
+
+	if epoch <= pau.latestEpoch {
+		return
+	}
+	ctx := context.Background()
+
+	resultStatus, err := pau.clientCtx.Client.Status(ctx)
+	if err != nil {
+		return
+	}
+	currentBlock := resultStatus.SyncInfo.LatestBlockHeight
+
+	response, err := pau.pairingQueryClient.Providers(ctx, &types.QueryProvidersRequest{
+		ChainID:    pau.rpcProviderEndpoint.ChainID,
+		ShowFrozen: true,
+	})
+	if err == nil && len(response.StakeEntry) > 0 {
+		for _, provider := range response.StakeEntry {
+			if !provider.IsAddressVaultOrProvider(provider.Address) {
```

Comment:
The check is good, the value is incorrect you will always hit true but also on providers which are not yours. 
you need to use your process public address that needs to be one of the values in the struct.

Reply:
Fixed in [5b112c0](https://github.com/lavanet/lava/pull/1729/commits/5b112c07493613be8c622cf5c60b360cacf8fd82)